alg/polygonize.cpp

```diff
@@ -664,9 +172,6 @@ static CPLErr GDALPolygonizeT(GDALRasterBandH hSrcBand,
         eErr = GDALRasterIO(hSrcBand, GF_Read, 0, iY, nXSize, 1, panThisLineVal,
                             nXSize, 1, eDT, 0, 0);
 
-        if (eErr == CE_None && hMaskBand != nullptr)
```

Comment:
Is the removal of the masking appropriate here ? Isn't there a risk of setting a different ID in the first and second pass if both don't apply the masking ?

Reply:
This is indeed a problem, it need to mask the raster data while reading.
Since the polygonizer needs a full labeling raster as its input(via Polygonizer::processLine), it means that any zero area defined by the mask should be assigned to a unique value, but this looks like a time-consuming job to find the unique value.

Reply:
Because the most used case is to use nodata area as mask, so the nodata value serve as the unique value, and all looks ok in that case.
And I think the logic of the polygonizer can be update to take into account the "invalid polygon" concept.

Reply:
Agreed that nodata is probably the main use case, but we should make sure that it works with an arbitrary mask band, not necessarily tied to nodata. I would suggest you restore the use of the mask band in the first pass, per this initial pull request, and potentially come with improvements/optimizations in a follow-up pull request. I don't think the main perf improvements of the new algorithm are related to that

Reply:
I've tried to fix this problem, it should work well but with little memory usage increasing because the polygonizer may hole the invalid polygon for a long time.

Reply:
do we need an extra test case in autotest to catch those cases?

Reply:
No, no problem. It is caused by my carelessness, actually the result is correct, I don't remember why I written the wrong feature count  in the benchmark table, now I have updated it.
Yes, we'd better add a none nodata mask as a test case, it would help us to avoid this situation in the feature.

Reply:
> Yes, we'd better add a none nodata mask as a test case

do you want to add that in this pull request ?

Reply:
Yes, just now I've add a test case.